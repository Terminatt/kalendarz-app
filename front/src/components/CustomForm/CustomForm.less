@import '~styles/variables.less';

.custom-form {
  .custom-form-btns {
    text-align: center;
    margin-top: @margin-big;

    .custom-form-btn {
      margin-right: @margin-small;
    }
  }
}