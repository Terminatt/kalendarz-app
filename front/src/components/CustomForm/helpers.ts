import { ResponseError } from '@generics/generics';
import { getEntries } from '@utils/general';

export function parseErrorResponse(errorResponse: ResponseError, ERROR_MESSAGES: {[key: string]: string}):
Array<{name: string, errors: string[]}> {
    const entries = getEntries(errorResponse);
    const errors: Array<{name: string, errors: string[]}> = [];

    // TODO SERVE ARRAY CASE
    entries.forEach((el) => {
        if (el[1] && !Array.isArray(el[1])) {
            errors.push({ name: el[0], errors: [ERROR_MESSAGES[el[1].type]] });
        }
    });

    return errors;
}
