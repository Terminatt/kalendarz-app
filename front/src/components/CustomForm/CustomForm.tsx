import { errorMessages, formLayout } from '@constants/constants';
import React, { useEffect } from 'react';
import { Form, FormProps } from 'antd';
import CustomButton from '@components/CustomButton/CustomButton';
import { GenericReactContent, ResponseError } from '@generics/generics';
import { parseErrorResponse } from './helpers';

import './CustomForm.less';

export interface CustomFormProps {
  formProps: Omit<FormProps, 'className'>;
  children?: GenericReactContent;
  className?: string;
  primaryBtnText: string;
  isLoading?: boolean;
  errorResponse?: ResponseError | null;
}

const CustomForm: React.FC<CustomFormProps> = (props) => {
    const {
        formProps, className, children, primaryBtnText, isLoading, errorResponse,
    } = props;
    const { form, ...restFormProps } = formProps;

    const clearForm = () => {
        form?.resetFields();
    };

    useEffect(() => {
        if (!errorResponse) {
            return;
        }
        const errors = parseErrorResponse(errorResponse, errorMessages);

        form?.setFields(errors);
    }, [errorResponse]);

    return (
        <Form className={`custom-form ${className || ''}`} form={form} {...restFormProps} {...formLayout}>
            {children}
            <div className="custom-form-btns">
                <CustomButton
                    disabled={isLoading}
                    loading={isLoading}
                    className="custom-form-btn"
                    htmlType="submit"
                >
                    {primaryBtnText}

                </CustomButton>
                <CustomButton className="custom-form-btn" variant="clear" onClick={clearForm}>Wyczyść</CustomButton>
            </div>
        </Form>
    );
};

export default CustomForm;
