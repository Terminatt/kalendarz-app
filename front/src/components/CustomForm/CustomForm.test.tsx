/* eslint-disable @typescript-eslint/no-floating-promises */
import GeneralUtils from '@utils/general';
import { Form, Input } from 'antd';
import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CustomForm from './CustomForm';

GeneralUtils.jestWorkAround();

test('Custom Form is rendered correctly', () => {
    const component = renderer.create(
        <CustomForm formProps={{}} primaryBtnText="Click">
            <Form.Item
                label="Nazwa użytkownika"
                name="username"
            >
                <Input placeholder="Podaj nazwę użytkownika" />
            </Form.Item>
        </CustomForm>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

// ant-form-item-explain-error
test('Primary Button text is rendered correctly', () => {
    const result = render(
        <CustomForm formProps={{}} primaryBtnText="Click">
            <Form.Item
                label="Nazwa użytkownika"
                name="username"
            >
                <Input placeholder="Podaj nazwę użytkownika" />
            </Form.Item>
        </CustomForm>,
    );
    const btn = result.container.querySelector('.custom-btn-primary');
    expect(btn).toHaveTextContent('Click');
});
