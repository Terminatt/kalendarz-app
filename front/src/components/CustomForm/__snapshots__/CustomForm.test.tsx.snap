// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Form is rendered correctly 1`] = `
<form
  className="ant-form ant-form-horizontal custom-form "
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className=""
        htmlFor="username"
        title="Nazwa użytkownika"
      >
        Nazwa użytkownika
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Podaj nazwę użytkownika"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="custom-form-btns"
  >
    <button
      className="ant-btn custom-btn custom-btn-primary custom-form-btn"
      onClick={[Function]}
      type="submit"
    >
      <span>
        Click
      </span>
    </button>
    <button
      className="ant-btn custom-btn custom-btn-clear custom-form-btn"
      onClick={[Function]}
      type="button"
    >
      <span>
        Wyczyść
      </span>
    </button>
  </div>
</form>
`;
